### ISSUE TITLE
Multi-consumption of an `AsyncSequence`

### ISSUE THREAD

**Issue by ABridoux** [apple/swift-async-algorithms#110]
Hi, I have been reading the documentation for this repository and it left me wondering: is this planned to introduce a feature to allow several tasks to consume the same `AsyncSequence`?

_If this is not the right place to offer ideas, please let me know. I don't think this would fit in a proposal._

## Context
I recently had to stream a file using `URLSession` and two pieces of code were interested in the streamed values. Rather than starting a stream twice which would not be efficient, I wanted to allow the stream to be consumed by several tasks. The data that were already streamed would be sent when a new consumption is set up.

## Basic implementation
I tried to implement such a solution (in a [gist](https://gist.github.com/ABridoux/913dfb95fabc22048b0a1c566059105a)) that will use a reduce function on the already emitted values and emit the result when a new task starts consuming the sequence. It's far from being perfect but I think it might help understanding the idea.

<details><summary>Implementation</summary>
<p>

```swift
struct ReducedReplayAsyncStream<Element> {

    typealias Reduce = (_ partialResult: inout Element, _ nextResult: Element) -> Void

    private let storage: _Storage
    private var originalStream: AsyncStream<Element>

    init(
        bufferingPolicy limit: AsyncStream<Element>.Continuation.BufferingPolicy = .unbounded,
        initialResult: Element,
        reduce: @escaping Reduce,
        build: (AsyncStream<Element>.Continuation) -> Void
    ) {
        originalStream = AsyncStream(Element.self, bufferingPolicy: limit, build)
        storage = _Storage(stored: initialResult, reduce: reduce)
    }

    private func makeStream() -> AsyncStream<Element> {
        AsyncStream<Element> { continuation in
            Task {
                var isFirst = false
                if await !storage.didStart {
                    await storage.setDidStart(true)
                    isFirst = true
                    startConsumingOriginalStream()
                }

                if !isFirst {
                    await continuation.yield(storage.stored)
                }
                await storage.appendContinuation(continuation)
            }
        }
    }

    private func startConsumingOriginalStream () {
        Task {
            for await value in originalStream {
                await storage.updateWith(value: value)
            }
            await storage.continuations.forEach { $0.finish() }
        }
    }
}

extension ReducedReplayAsyncStream {

    private actor _Storage {
        private let reduce: ReducedReplayAsyncStream.Reduce

        var didStart = false
        var stored: Element
        var continuations: [AsyncStream<Element>.Continuation] = []

        init(stored: Element, reduce: @escaping Reduce) {
            self.stored = stored
            self.reduce = reduce
        }

        func updateWith(value: Element) {
            reduce(&stored, value)
            continuations.forEach { $0.yield(value) }
        }

        func setDidStart(_ value: Bool) {
            didStart = value
        }

        func appendContinuation(_ continuation: AsyncStream<Element>.Continuation) {
            continuations.append(continuation)
        }
    }
}

extension ReducedReplayAsyncStream: AsyncSequence {
    typealias AsyncIterator = AsyncStream<Element>.AsyncIterator

    func makeAsyncIterator() -> AsyncIterator {
        let stream = makeStream()
        return stream.makeAsyncIterator()
    }
}
```

</p>
</details>

<details><summary>Usage</summary>
<p>

```swift
var subscriptions: Set<AnyCancellable> = []
var continuation: Stream.Continuation!

let replayStream = ReducedReplayAsyncStream<Int>(
    initialResult: 0,
    reduce: { partialResult, nextResult in partialResult = partialResult + nextResult },
    build: { continuation = $0 }
)

var counter = 0
Timer.publish(every: 0.4, on: .main, in: .default)
    .autoconnect()
    .sink { _ in
        if counter == 10 {
            continuation.finish()
        }
        continuation.yield(counter)
        counter += 1
    }
    .store(in: &subscriptions)

Task {
    for await value in replayStream {
        print("[A]", value)
    }
}

Task {
    try await Task.sleep(nanoseconds: 3_000_000_000)
    for await value in replayStream {
        print("[B]", value)
    }
}
```

</p>
</details>

Some considerations about efficiency can be found in the gist.

**Comment by BrentMifsud** [comment]
Theres already an open source library that is attempting to do this. I definitely think that "official" support of this would be ideal. Maybe a collaboration with the owner of this repo:

https://github.com/AsyncCommunity/AsyncExtensions

**Comment by justin-foreflight** [comment]
Here's ([gist](https://gist.github.com/justin-foreflight/0f896325a51b28edc1a743751cc3598b)) my not very Swifty attempt, main thing I was going for was cancelling all the other streams would cancel the original.

**Comment by phausler** [comment]
I think solutions in this territory are interesting; The one issue is that more often than not the buffer may be indefinite/unbounded.

I think the uses are perhaps the important point here - I don't think that it is that folks need to replay all values in very distant parts of code execution but instead it feels like there are code locations that have short execution distances from each other that would want to use the values at the same time. 

So:

```
func doThings<Input: AsyncSequence>(_ input: Input) {
  let (sideA, sideB) = input.split() 
  let a = sideA.compactMap { 
    switch $0 {
      case .a(let payload): return payload
      default: return nil
    }
  }
  let b = sideb.compactMap { 
    switch $0 {
      case .b(let payload): return payload
      default: return nil
    }
  }
  Task {
     await doThingWithA(a)
   }
  Task {
     await doThingWithB(b)
   }
}
```

**Comment by ABridoux** [comment]
I believe you are right, I cannot think of a use case where all values should be replayed. The feature I wanted to talk about was rather to "jump in the wagon" and start receiving the remaining values of the stream without impacting other consumers. The reduce function would be used for that: to get the "synthesis" of all the previous values before receiving the remaining ones.
I think this could be useful when it's not possible to make a new sequence. For instance when it's too expensive like a stream of a large file. Though I agree that this might be too specific to fit in this library.

**Comment by phausler** [comment]
a stream from a large file should be pretty cheap; `AsyncBufferedByteIterator` is quite efficient and that type of strategy should be used to stream bytes.

If the rate of one consumer is less than another then there needs to be some sort of buffering or back pressure to accommodate for the slowest consumer. Else-wise there will be dropped values. The problem is knowing the number of consumers ahead of time. If you can know that then you can make back pressure work.

**Comment by ABridoux** [comment]
I have been thinking about the function `AsyncSequence.split()` you imagined in your first response @phausler and I think I misunderstood how `AsyncSequence`s were supposed to be used.

A `Publisher` in Combine can have several `Subscrriber`s.. This is useful if several parts of the code want to observe the same stream of values. Like for instance, observe whether the app can access the internet or not. For this solution with Combine, I would offer a static value on an enum like

```swift
enum Reachability {
    static let isOnline: AnyPublisher<Bool, Never> {
        // here a `NWPathMonitor` is instantiated to create the publisher
    }()
```

The benefit of this approach is to avoid re-creating a `NWPathMonitor` each time a piece of code wants to observe the online status. The example could also work when it's heavy to build the publisher.

Now this approach doesn't work with `AsyncSteam` (assertion failure) or `AsyncChannel` (at least from my experimentation with it). So the approach here would rather be to create a new `AsyncStream` or `AsyncChannel` each time the property is computed. Something like that:

```swift
enum Reachability {

    static var isOnlineValues: AsyncStream<Bool, Never> {
        AsyncStream<Bool> { continuation in
            let monitor = NWPathMonitor()
            continuation.onTermination = { @Sendable [monitor] _ in monitor.cancel() }
            monitor.pathUpdateHandler = { path in
                let isOnline = MonitorWrapper.isStatusOnline(status: path.status)
                continuation.yield(isOnline)
            }
            monitor.start(queue: .main)
        }
    }
}
```

So my question is: is this the right way of thinking about `AsyncSequence`? Is this the right approach to create a new sequence each time a new piece of code needs to consume the values? (I wonder if it's the way `Publisher.values` works). And if it's the right approach, then aren't there use cases when it might be less efficient to create a new sequence rather than use the existing one?

**Comment by ABridoux** [comment]
@phausler I see that the multi-consumption was already discussed in a thread about `AsyncSequence`. Especially I wanted to share your words from [a response](https://forums.swift.org/t/should-asyncsequence-replace-combine-in-the-future-or-should-they-coexist/53370/10) here before asking my question:

> Distributors:
> Sending a value to an AsyncSequence where the send awaits the consumption.
> 
> Sharing the iteration in a safe manner across multiple consumers where each consumer gets a distinct value
> 
> Sharing the values from iteration among a known number of consumers

That's exactly what I meant in the first place (sorry if that was unclear). A way for multiple consumers to receive values emitted by a single sequence. Very much like a publisher-subscribers relationship.

Is this this planned to add such a feature? When I experimented with `AsyncChannel` it was not possible to consume the sequence in two different places:
```swift
func testAsyncChannel() {
    let channel = AsyncChannel<Int>()

    Task {
        await channel.send(1)
        await channel.send(2)
    }

    Task {
        for await value in channel {
            print("A", value)
        }
    }

    Task {
        for await value in channel {
            print("B", value)
        }
    }
}
```
```swift
// prints:
// A 1
// B 2
```

I have got some uses cases were it would be useful to have this logic available. For instance with a re-implementation of `@Published`. Maybe I am biased by my usage of Combine, but I think this could be really useful to have such a property wrapper that will emit modifications of a value though an `AsyncSequence`.
The only solution I can think of currently would be to create a new `AsyncStream` each time a consumer is interested and to store the associated continuations to forward the emitted values of the original stream.

**Comment by phausler** [comment]
There are some unanswered questions with regards to sharing behaviors. For example; is it the best route to go having it be an "operator" style API? Combine takes the approach of multicast+autoconnect, does that make sense? or does some other composed behavior split make more sense? How can we make the api more approachable so that it is clear on when you need it? The list goes on.

Personally I think the multicast+autoconnect is a bit complex of a concept, and instead breaking it up into concepts of sharing iteration or sharing values seems to me more approachable if you are not experienced with functional reactive programming. 

I also contend that splitting into a known number of replications can allow us to make some really nice behavioral aspects with regards to cancellation.

The general role that share plays to me seems quite useful and needed, just not the forefront of need.

The property wrapper thing brings up a ton of other questions; namely of which are more aimed at groups working at Apple on SwiftUI to answer. Since in truth, even though `@Published` may live in Combine, it is really more of a SwiftUI type in practice.

**Comment by ABridoux** [comment]
I understand now that `AsyncSequence` approach is quite different from the Combine publishers. All your points make sense.
I am left wondering if the best solution for now is to create a new sequence each time a consumer is interested. This might be expensive in some scenarios when creating a new sequence require some heavy work.

Regarding the `@Published` to be honest I use it in many other places than with SwiftUI. Especially, I find it useful for testing purposes to easily observe if a value has changed or a function was called with the proper input. But it might not be the right approach with `AsyncSequence` indeed (I'd rather use something like a custom `@AsyncNext` property wrapper to observe the next value).

Thank you a lot for taking the time to answer my questions. If that's ok I'll close the issue in a few days if no comments are added.

**Comment by phausler** [comment]
Perhaps the parts that we can help identify here are: what pieces do we need to build such a thing? And what are it's restrictions?

First and foremost: any such property wrapper would have a couple of characteristics. 
It would effectively always be a trigger on `didSet` because it is awaited.
The setter currently cannot be async so we need to address back pressure (e.g. what happens if no one is currently awaiting a value?)
To the point of the last question: should it even be settable beyond the first value?
What happens if more than one interested party is awaiting?
What does the projection of the property do? is it just an `AsyncSequence`?

Out of those questions; what parts can we build independently to make constructing that thing easy/safe?
Just because it incubates in this project does not mean that it must always stay here - perhaps something of useful importance to the rest of the language might be able to transition into the swift concurrency library.

**Comment by phausler** [comment]
One idea that I had a bit ago is to do something like this:

```swift
@propertyWrapper
public struct ThisDeservesABetterName<Element> {
  public private(set) var wrappedValue: Element
  let channel = AsyncChannel<Element>()
  
  public init(wrappedValue: Element) {
    self.wrappedValue = wrappedValue
  }
  
  public struct Projection: AsyncSequence {
    public struct Iterator: AsyncIteratorProtocol {
      var iterator: AsyncChannel<Element>.Iterator
      
      public mutating func next() async -> Element? {
        return await iterator.next()
      }
    }
    
    var value: Element
    let channel: AsyncChannel<Element>
    
    public mutating func setValue(_ value: Element) async {
      self.value = value
      await channel.send(value)
    }
    
    public func makeAsyncIterator() -> Iterator {
      Iterator(iterator: channel.makeAsyncIterator())
    }
  }
  
  public var projectedValue: Projection {
    get {
      Projection(value: wrappedValue, channel: channel)
    }
    set {
      wrappedValue = newValue.value
    }
  }
}
```

**Comment by malhal** [comment]
I found this [SharedAsyncSequence](https://github.com/reddavis/Asynchrone#sharedasyncsequence) via [Consuming an AsyncStream from multiple Tasks](https://forums.swift.org/t/consuming-an-asyncstream-from-multiple-tasks/54453)

**Comment by phausler** [comment]
I think there are some expectations that may not meet.

It should use back pressure not buffering: e.g. the next item should await all consumption of previous next calls to be done.

It should rethrow failures if and only if the base throws.

**Comment by phausler** [comment]
I agree that this is an important feature to iterate on (pun most assuredly intended). However the battle that we must face is that `share` in Combine feels a bit abject and mysterious some times. Some folks find it hard to understand why/when it is needed. 

If we can somehow leverage things in the language that help to that end it would go a long way to making it more of a concrete item. Even though personally I find a share or split operator concept for `AsyncSequence` meaningful and useful, other folks are not yet convinced.

**Comment by malhal** [comment]
If you turn the problem on it's head (like using value types vs objects) then you can have multiple clients using the continuation and you only need a single consumer of the stream. Also you don't need to worry about the re-entrency issue of async funcs or type erasure.

**Comment by KeithBauerANZ** [comment]
Was trying to implement one of these behaviors myself (one sequence's elements consumed once each by multiple concurrent consumers) [in this thread](https://forums.swift.org/t/crash-in-safe-swift-with-actors-concurrency/64496) and got sent here to make a feature request, so

+1

**Comment by mantuness** [comment]
+1 on this

**Comment by nonameplum** [comment]
+1

**Comment by malhal** [comment]
Just make a channel per client, have the server loop the main sequence and send to all client channels.

**Comment by dylanjrafferty** [comment]
Checking back for any updates on this

**Comment by fawkeswei** [comment]
+1

**Comment by kbista001** [comment]
+1

**Comment by bok-** [comment]
Looks like Swift 6.1 will fix the issues that `AsyncStream` has with propagating termination: https://github.com/swiftlang/swift/pull/75878

And tests supporting multi-consumers in https://github.com/swiftlang/swift/pull/76571 though perhaps not officially documented support.

Definitely owe @jamieQ a drink!

**Comment by mattmassicotte** [comment]
Ok I knew that @jamieQ rocked, but I’m not sure I realized just how hard.

**Comment by adam-zethraeus** [comment]
@jamieQ — thank you.

**Comment by jamieQ** [comment]
thanks, but just to be clear – i definitely did _not_ implement the core motivation behind this issue, at least as i understand it (supporting multiple stream consumers that 'see' each element). i just fixed a termination bug and have an open PR to try and make the two stream implementations more consistent.

**Comment by djs66256** [comment]
+1

**Comment by wouterbrink** [comment]
I'm looking exactly for this use case. Does anyone know if there are any plans to implement this or if the idea was discarded?

**Comment by esnssr** [comment]
jumping on to this conversation. it feels like this should have been an early feature with AsyncSequence. regardless, is there any proposal or PR for when this is coming?
whether it’s in SwiftLang or here?

**Comment by malhal** [comment]
My theory is it only feels that way to those new to Swift's value semantics which like the inverse of the more familiar reference semantics. So requires solving problems in a kinda of backwards way and multi consumption of stream values isn't needed or even is possible.

**Comment by mattmassicotte** [comment]
I think this thread makes it very clear this feature is highly-desired. In my experience, it is frequently encountered as a major hurdle when trying to migrate away from Combine.

If you believe there is an alternative approach that matches or exceeds the ergonomics, sufficiently solves the problem, and/or is more idiomatic, I think everyone listening in would greatly appreciate more details.

**Comment by dylanjrafferty** [comment]
@malhal While this in essence is the solution here and building your own broadcasting mechanism (like this array you outlined here) works- I think that a solution within the async-algorithms or stdlib itself needs to exist. This is a super essential piece (in my opinion, and also in pretty clearly of many others with proof being this thread) of infrastructure that you need to build meaningful reactive streams similar to to what Combine and RxSwift support. 

Without a solution and creation of a new AsyncChannel type that supports it OR a modifier that enables that behavior across AsyncSequences or streams- it leaves you without a major tool in a reactive system. I personally have found that I had to choose between rolling my own solution EVERY single time I wanted to create a stream and observe it in multiple places OR creating my own type that supports it and deviating from the stdlb/async-algorithms offering.

**Comment by BrentMifsud** [comment]
@dylanjrafferty 

Pretty much my experience as well.

I always end up having to build something like this:

```swift
actor SharedAsyncStream<T> {
    private var continuations = [UUID: AsyncStream<T>.Continuation]()

    var values: AsyncStream<T> {
        let uuid = UUID()
        let (stream, continuation) = AsyncStream<T>.makeStream()
        continuation.onTermination = { _ in 
            Task {
                 await removeContinuation(for: uuid)
            }
        }
        
        continuations[uuid] = continuation
        return stream
    }

    func yield(value: T) {
        for (_, continuation) in continuations {
            continuation.yield(value)
        }
    }

    private func removeContinuation(for uuid: UUID) {
        _ = continuations.removeValue(forKey: uuid)
    }
}
```

**Comment by malhal** [comment]
That `SharedAsyncStream` is kind of depressing, i.e. misuse of `AsyncStream` (it should own the event producing object), misuse of continuations (those are for interfacing with old callback code), caused a retain cycle (onTermination retains self to call removeContinuation), broken co-operative cancellation (you should not create new top level Tasks to do low level work).

**Comment by adam-zethraeus** [comment]
> misuse of AsyncStream (it should own the event producing object),
Would you be able to point me to a source for this assertion?

> misuse of continuations (those are for interfacing with old callback code)
"A mechanism to interface between synchronous and asynchronous code, logging correctness violations." Why old or callbacks?

**Comment by mattmassicotte** [comment]
I would love to see a resolution to this issue, and the sooner the better. But in the mean time, sharing potential ideas and workarounds is good and very welcome. I don't think it is an appropriate to criticize those implementations.

But, if you feel compelled to do so, I think it should be done with both kindness and respect.

**Comment by dylanjrafferty** [comment]
I would be more than happy to contribute to this and actually implement it myself, but I don't feel good about going in blind here since there are obviously some considerations by the team at apple that we aren't privy too in play here. I am glad to see this thread picking up some traction again, maybe this time can get us closer to a resolution!

**Comment by Sajjon** [comment]
More importantly, @phausler wrote [this (Swift Forums)](https://forums.swift.org/t/pitch-transactional-observation-of-values/78315/14):

> Indirectly yes? I it is the first of its' kind and one could potentially build some things around Observed I would imagine.

As a reply to [Kaans observation](https://forums.swift.org/t/pitch-transactional-observation-of-values/78315/8) that the pitch does use multiple consumers for the single producer

> Does this mean that this pitch will also enable single-producer multi-consumer for all AsyncSequences?

So that gives some hope...

**Comment by tcldr** [comment]
I'm going to repeat my thoughts [from the Swift forums](https://forums.swift.org/t/swift-async-algorithms-proposal-broadcast-previously-shared/61210/52) regarding this algorithm:

--- 

I’ll jump in here to say I haven’t been actively developing my own ideas on this further, but I still firmly believe that the ideal solution to this looks quite a bit different to those above.

The crux of it is that the range of problems that this algorithm should attempt to solve is broader than a single algorithm.

Firstly, as you begin designing a multicast algorithm such as the one above, it becomes clear that there are many variations on a 'broadcast' algorithm, and therefore it's important that we design a solution to accommodate future enhancements.

The following comment by [@ktoso](https://forums.swift.org/u/ktoso) summarises some of these:

> How does that look like in practice? Explicit operators which define the various semantics, like:
> 
>  - cancellation:
>    -  cancel when one cancels
>    - cancel when all cancel etc
>  - buffering / replay / live:
>    - buffer and replay to all subscribers
>    - broadcast from current element to all subscribers
>    - round robin elements to active subscribers" behaviors
>  - when elements are emitted
>  - when completion is signaled

Secondly, and perhaps more contentiously, I'm still unclear as to why this family of broadcast algorithms is expected to be chained (like map or filter) rather than being source algorithms (like AsyncStream or Subject's in the Rx/Combine world.)

The argument that I've heard to date is that programmers will find the 'chaining' algorithms simpler to understand as they struggle with the concept of Subjects in Rx/Combine.

However, I think that this is to miss some of the benefits of this model.

One of the big things that I think new programmers struggle with is that they are unable to iterate over source sequences (such as AsyncStream) multiple times – they're 'single shot'. Concerningly, the behaviour when you do is undefined.

If we were to instead adopt the source algorithm approach to multicasting, we would benefit from not only a range of multicasting options, but also a range of source algorithms that can safely be iterated multiple times. (And also designed to accommodate other issues such as back pressure.)

Of course, this approach would take some design effort, likely the introduction of a new AsyncSource type (roughly equivalent to a Subject) in the stdlib and then a new chained operator to multicast via said AsyncSource types. But I think it would go a long way to solving many of the shortcomings we have with AsyncSequence today.

--- 

I still believe this to be the case, as it ties in with Swift's progressive disclosure philosophy. Especially as 'single shot', performance focused subjects can still be trivially implemented if that's a necessity for someone's application.

Also, since I posted, there seems to have been a few changes/proposed changes to the concurrency model that would have a bearing on the design:

  - Previously it was quite difficult to define which executor the underlying 'driver' Task would run, I'm guessing that should be a bit easier now. One of the unfortunate issues that asynchronous sequences use to have was forced actor-hopping due to many of its methods being non-isolated. With actor inheritance I'm hoping this will be less of an issue.
  - In addition, I'm hoping that the [starting tasks synchronously from caller context](https://forums.swift.org/t/pitch-concurrency-starting-tasks-synchronously-from-caller-context/77960) pitch might make some of the replay/behavior subject like properties a possibility, too. Without this, you can get into unfortunate race conditions without careful management.

**Comment by malhal** [comment]
In an attempt to generalise some previous attempts I've mocked up a 'with' style `func withBroadcast` that could be used as follows:

```
// example source stream
let stream = AsyncStream<Int> { continuation in
    Task {
        for i in 1...5 {
            continuation.yield(i)
            try? await Task.sleep(for: .seconds(1))
        }
        continuation.finish()
    }
}

await withBroadcast(from: stream) { broadcast in
    // async

    broadcast.addChannel { channel in
        for await value in channel {
            print("Channel 1 received: \(value)")
        }
        print("Channel 1 ended")
    }

    try? await Task.sleep(for: .seconds(1)) // simulate delay in dynamically adding channel

    broadcast.addChannel { channel in
        for await value in channel {
            print("Channel 2 received: \(value)")
        }
        print("Channel 2 ended")
    }
}

print("Broadcasting ended")
```

The idea came to me because I realised we essentially need a wrapper around `withTaskGroup` to start the source stream so cancellation can be propagated. `addChannel` would create a `Task` that creates an `AsyncChannel` and adds it to an actor `ChannelManager`, callbacks the `channel in` closure (where the `for await value in channel` is), then upon that for ending, control passes back to the `Task` where it finishes and removes the channel from the actor. Also I believe that if the source stream throws, the actor (which has lifetime tied to task group) would deinit and it could remove and finish all active channels.

**Comment by phausler** [comment]
that is an interesting idea for the `with*` style api - It makes me wonder if we can use ~Escapable to do that then?

**Comment by FranzBusch** [comment]
> that is an interesting idea for the `with*` style api - It makes me wonder if we can use ~Escapable to do that then

I don't think we can use `~Escapable` for this if I understand the proposal correctly here. The `with-style` API here is not only for lifetime tracking but to hide a `withTaskGroup` API under the hood to setup a child task to iterate the upstream. So as long as `TaskGroup` doesn't provide a `non-with-style` we can't avoid it here.

Overall I agree this is an interesting idea!

**Comment by tcldr** [comment]
> Overall I agree this is an interesting idea!

This is an interesting idea, and I think it's a fine utility for certain use cases.

However, for a general purpose algorithm I think we have at least one major drawback that could be solved with a different design: AFAICT it requires knowing all the destinations up front as the API is suspending. My feeling is a general purpose multi-consumption sequence is going to need to allow consumers to come and go.

Also, as we can't use `~Escapable` the source sequence can still be re-iterated. This isn't much worse than the current situation though, and perhaps as the API is suspending the source sequence would have been cancelled prior to the caller being resumed, preventing the caller from handing-off the sequence elsewhere. (I'm assuming a cancelled one-shot sequence is safer than a running one, here.)

**Comment by malhal** [comment]
Please take another look. The destinations are dynamic.

**Comment by malhal** [comment]
Is the idea for ~Escapable just to be more performant than an actor for the client channels?

**Comment by tcldr** [comment]
> Please take another look. The destinations are dynamic.

I'm probably missing something, but how can consumers come along outside of the `withBroadcast` block?

**Comment by malhal** [comment]
Ah ok I think you're saying you'd like different top level Tasks (e.g. defined in different parts of the app) to do for await in a shared stream.

**Comment by tcldr** [comment]
> Ah ok I think you're saying you'd like different top level Tasks (e.g. defined in different parts of the app) to do for await in a shared stream.

Yes, exactly. I often find myself wanting to expose a shared stream on a type that consumers can iterate on without needing to know about who those consumers are.

**Comment by phausler** [comment]
> The `with-style` API here is not only for lifetime tracking but to hide a `withTaskGroup` API under the hood to setup a child task to iterate the upstream.

I was suggesting that perhaps `withTaskGroup` should be folded inside out with `~Escapable`. But moreover the fundamental issue with sharing is that it has been blocked on the same reason that withTaskGroup is a closure currently.

To be clear; I am not saying we should predicate blocking exploring that idea on `withTaskGroup` supporting `~Escapable`; my guess is that is a long way off. 

If someone wants to take the time to prototype the with* style variation; I think it would be worthwhile to at least consider. Worst case scenario; it gives us some insight on the possiblities. I think it is a fine shape for a general use algorithm for how it is possible today.



> I often find myself wanting to expose a shared stream on a type that consumers can iterate on without needing to know about who those consumers are.

That does not preclude the cases where the consumers are known; I have a feeling that there is a few variants of these types of sharing/broadcasting. One style you have a fixed and known set of consumers that you want to ensure all consumers are getting the same values at the same iteration. Another type is that you want all consumers to get the latest of the values and none of them are interdependent upon waiting on each other.  Those are just two variants on the concept, I could easily envision a few more.

**Comment by Sajjon** [comment]
P Hausler wrote a proposal of "share" algorithm yesterday:

https://forums.swift.org/t/kickoff-of-a-new-season-of-development-for-asyncalgorithms-share/81447
